### PR TITLE
Add roundtrip ckpt conversion tests in CI

### PR DESCRIPTION
**Pull request by hengtaoguo**
# Description
- Make `import torch` lazy import in `to_maxtext`, so we can run torch-free conversion in CI.
- Add a 2-layered qwen3-30b-a3b ckpt-conversion-roundtrip cpu-only test in CI. 
-- We lazy-load safetensors so only 2-out-of-16 chunks were actually loaded for `to_maxtext` conversion. 
-- Within the tmpdir, we convert maxtext Orbax ckpt back to HF for `to_huggingface` conversion.
-- All converted MaxText/HF ckpts will be auto removed from tmpdir.
-- Success criteria: (1) Subprocess finishes without errors (2) Result ckpts exist in the tmpdir. We didn't enforce logits check in the CI because that will actually require torch dependency. But I've tested forward_pass_logit_checker offline.
- Fix [b/513636729](https://b.corp.google.com/issues/513636729), introduced in [cl/914979909](https://critique.corp.google.com/cl/914979909). 

# Tests

```
# to maxtext
python -m maxtext.checkpoint_conversion.to_maxtext src/maxtext/configs/base.yml model_name=qwen3-30b-a3b base_output_directory=/dev/shm/mini base_num_decoder_layers=2 override_model_config=True scan_layers=false hf_access_token=\<token\> hardware=cpu skip_jax_distributed_system=True checkpoint_storage_use_ocdbt=False checkpoint_storage_use_zarr3=False --eager_load_method=safetensors --save_dtype=bfloat16 --lazy_load_tensors=True

# to huggingface
python -m maxtext.checkpoint_conversion.to_huggingface model_name=qwen3-30b-a3b load_parameters_path=/dev/shm/mini/0/items base_output_directory=/dev/shm/mini/hf_safetensor/qwen3-30b-a3b base_num_decoder_layers=2 override_model_config=True scan_layers=false weight_dtype=bfloat16 hardware=cpu skip_jax_distributed_system=True hf_access_token=\<token\> --override_model_architecture=True

# logits check (roundtrip above)
python -m tests.utils.forward_pass_logit_checker maxtext/configs/base.yml run_name=ht_test model_name=qwen3-30b-a3b tokenizer_path=Qwen/Qwen3-30B-A3B-Thinking-2507 load_parameters_path=/dev/shm/mini/0/items base_num_decoder_layers=2 override_model_config=True scan_layers=false max_prefill_predict_length=4 ici_tensor_parallelism=4 hf_access_token=\<token\> --run_hf_model=True --hf_model_path=Qwen/Qwen3-30B-A3B-Thinking-2507 --hf_model_path=/dev/shm/mini/hf_safetensor/qwen3-30b-a3b --max_kl_div=0.015
```

2-layerd mini checkpoint with matching gibberish logits: [Logs](https://paste.googleplex.com/5895969509081088)

Tests active in CI: https://screenshot.googleplex.com/4ZhRPiqJZLxTXk5

For this single test on my v5p-8 VM, it took ~2.5 minutes and the peak memory was ~12G.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

**Review by SurbhiJainUSC — APPROVED** [review]

File: tests/integration/checkpoint_conversion_test.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+          sys.executable,
+          "-m",
+          "maxtext.checkpoint_conversion.to_maxtext",
+          "src/maxtext/configs/base.yml",
+          f"model_name={model_name}",
+          f"base_output_directory={tmpdir}",
+          f"base_num_decoder_layers={base_num_decoder_layers}",
+          "override_model_config=True",
+          "scan_layers=False",
+          "weight_dtype=bfloat16",
```
Comment (shuningjin) [inline tests/integration/checkpoint_conversion_test.py]: `to_maxtext` is controlled by `--save_dtype=bfloat16` instead `weight_dtype=bfloat16`. https://github.com/AI-Hypercomputer/maxtext/blob/e6fcf3acc7db9c238d02a678ac6f2ffc9ac094f1/src/maxtext/checkpoint_conversion/to_maxtext.py#L1104

(`to_huggingface` is controlled by `weight_dtype=bfloat16`; we could resolve this inconsistency later)

File: tests/integration/checkpoint_conversion_test.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+          "hardware=cpu",
+          "skip_jax_distributed_system=True",
+          "checkpoint_storage_use_ocdbt=False",
+          "checkpoint_storage_use_zarr3=False",
+          "--lazy_load_tensors=True",
```
Comment (shuningjin) [inline tests/integration/checkpoint_conversion_test.py:49]: I guess by using `lazy_load_tensors`
- download only the safetensor for 2 layers instead of entire model?
- bypass torch, as it uses `safe_open(..., framework="np")`?

File: tests/integration/checkpoint_conversion_test.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+  @pytest.mark.cpu_only
+  def test_qwen3_30b_a3b_roundtrip_conversion(self):
+    model_name = "qwen3-30b-a3b"
+    base_num_decoder_layers = 2
+    with tempfile.TemporaryDirectory() as tmpdir:
+      cmd = [
+          sys.executable,
+          "-m",
+          "maxtext.checkpoint_conversion.to_maxtext",
+          "src/maxtext/configs/base.yml",
```
Comment (shuningjin) [inline tests/integration/checkpoint_conversion_test.py]: nit: remove for style consistency with to_huggingface cmd

File: src/maxtext/configs/pyconfig.py [hunk 4]
```diff
@@ -295,7 +295,10 @@ def __getattr__(self, attr: str) -> Any:
     # This is necessary for proper pickling/unpickling support
     flat_config = object.__getattribute__(self, "_flat_config")
     if attr in flat_config:
-      return flat_config[attr]
+      val = flat_config[attr]
+      if isinstance(val, dict) and attr in ("debug", "rl", "lora"):
+        return getattr(object.__getattribute__(self, "_pydantic_config"), attr)
+      return val
```
Comment (shuningjin) [inline src/maxtext/configs/pyconfig.py:301]: just curious: why is this change needed?

File: src/maxtext/checkpoint_conversion/to_maxtext.py [hunk 5]
```diff
@@ -79,6 +79,21 @@
 from safetensors import safe_open
```
Comment (shuningjin) [inline src/maxtext/checkpoint_conversion/to_maxtext.py:79]: nit: Could you update the following comment to reflect that 'safetensors' is the default eager load method instead 'transformers'

https://github.com/AI-Hypercomputer/maxtext/blob/e6fcf3acc7db9c238d02a678ac6f2ffc9ac094f1/src/maxtext/checkpoint_conversion/to_maxtext.py#L912-L914

File: tests/integration/checkpoint_conversion_test.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+          f"model_name={model_name}",
+          f"load_parameters_path={expected_dir}",
+          f"base_output_directory={expected_hf_dir}",
+          f"base_num_decoder_layers={base_num_decoder_layers}",
+          "override_model_config=True",
+          "scan_layers=false",
+          "weight_dtype=bfloat16",
+          "hardware=cpu",
+          "skip_jax_distributed_system=True",
+          "--override_model_architecture=True",
```
Comment (shuningjin) [inline tests/integration/checkpoint_conversion_test.py:80]: unrelated to this pr, for follow-up: I wonder if `--override_model_architecture=True` can be directly replaced with `override_model_config=True`

**Review by shuningjin — APPROVED** [review]
Looks good! Some minor comments.

File: src/maxtext/checkpoint_conversion/to_maxtext.py (hunk 5 above)
Comment (hengtaoguo) [reply src/maxtext/checkpoint_conversion/to_maxtext.py:79]: Nice catch and done.

File: src/maxtext/configs/pyconfig.py (hunk 4 above)
Comment (hengtaoguo) [reply src/maxtext/configs/pyconfig.py:301]: It is actually fixing [b/513636729](https://b.corp.google.com/issues/513636729). This change prevents the configuration wrapper from accidentally turning nested setting groups (like lora) into basic Python dictionaries, ensuring that scripts can still access sub-settings using standard dot notation like `config.lora.lora_restore_path`.

File: tests/integration/checkpoint_conversion_test.py (hunk 6 above)
Comment (hengtaoguo) [reply tests/integration/checkpoint_conversion_test.py:80]: IIUC `override_model_config` is for overriding MaxText models, while `override_model_architecture` is for overriding/dumping HF config.json. I also agree we should consolidate this flag in a follow up PR.

File: tests/integration/checkpoint_conversion_test.py (hunk 3 above)
Comment (hengtaoguo) [reply tests/integration/checkpoint_conversion_test.py]: Done

File: tests/integration/checkpoint_conversion_test.py (hunk 1 above)
Comment (hengtaoguo) [reply tests/integration/checkpoint_conversion_test.py]: Thanks for pointing this out! Updated.

File: tests/integration/checkpoint_conversion_test.py (hunk 2 above)
Comment (hengtaoguo) [reply tests/integration/checkpoint_conversion_test.py:49]: Yes. When we use lazy_load_tensors, it only requests the parameter keys for the first 2 layers, so it only download those safetensors chunks (2 out of 16 chunks) which contain the requested keys. And `framework="np"` is doing the torch-free trick here.
